test/performance/odp_ipsec.c

```diff
@@ -0,0 +1,1035 @@
+
+	param.proto = ODP_IPSEC_ESP;
+	param.dir = ODP_IPSEC_DIR_OUTBOUND;
+
+	if (cargs->tunnel) {
+		uint32_t src = IPV4ADDR(10, 0, 111, 2);
+		uint32_t dst = IPV4ADDR(10, 0, 222, 2);
+		odp_ipsec_tunnel_param_t tunnel;
+
+		memset(&tunnel, 0, sizeof(tunnel));
+		tunnel.type = ODP_IPSEC_TUNNEL_IPV4;
+		tunnel.ipv4.src_addr = &src;
+		tunnel.ipv4.dst_addr = &dst;
+		tunnel.ipv4.ttl = 64;
+
+		param.mode = ODP_IPSEC_MODE_TUNNEL;
+		param.outbound.tunnel = tunnel;
+	} else {
+		param.mode = ODP_IPSEC_MODE_TRANSPORT;
+	}
+
+	if (cargs->schedule || cargs->poll) {
+		out_queue = odp_queue_lookup("ipsec-out");
+		if (out_queue == ODP_QUEUE_INVALID) {
+			app_err("ipsec-out queue not found\n");
+			return ODP_IPSEC_SA_INVALID;
+		}
+		param.dest_queue = out_queue;
+	} else {
+		param.dest_queue = ODP_QUEUE_INVALID;
+	}
+
+	return odp_ipsec_sa_create(&param);
+}
+
+static uint8_t test_data[] = {
+	/* IP */
+	0x45, 0x00, 0x00, 0x80, 0x00, 0x00, 0x00, 0x00,
+	0x40, 0x01, 0xac, 0x27, 0xc0, 0xa8, 0x6f, 0x02,
+	0xc0, 0xa8, 0xde, 0x02,
+
+	/* ICMP */
+	0x08, 0x00, 0xfb, 0x37, 0x12, 0x34, 0x00, 0x00
+};
+
+static odp_packet_t
+make_packet(odp_pool_t pkt_pool, unsigned int payload_length)
+{
+	odp_packet_t pkt;
+
+	if (payload_length < sizeof(test_data))
+		return ODP_PACKET_INVALID;
+
+	pkt = odp_packet_alloc(pkt_pool, payload_length);
+	if (pkt == ODP_PACKET_INVALID) {
+		app_err("failed to allocate buffer\n");
+		return pkt;
+	}
+
+	odp_packet_copy_from_mem(pkt, 0, sizeof(test_data), test_data);
+	odp_packet_l3_offset_set(pkt, 0);
+
+	uint8_t *mem = odp_packet_data(pkt);
+	((odph_ipv4hdr_t *)mem)->tot_len = odp_cpu_to_be_16(payload_length);
+	memset(mem + sizeof(test_data), 1, payload_length - sizeof(test_data));
+
+	return pkt;
+}
+
+/**
+ * Run measurement iterations for given config and payload size.
+ * Result of run returned in 'result' out parameter.
+ */
+static int
+run_measure_one(ipsec_args_t *cargs,
+		ipsec_alg_config_t *config ODP_UNUSED,
+		odp_ipsec_sa_t sa,
+		unsigned int payload_length,
+		crypto_run_result_t *result)
+{
+	odp_ipsec_out_param_t param;
+	odp_pool_t pkt_pool;
+	odp_queue_t out_queue;
+	odp_packet_t pkt = ODP_PACKET_INVALID;
+	int rc = 0;
+
+	pkt_pool = odp_pool_lookup("packet_pool");
+	if (pkt_pool == ODP_POOL_INVALID) {
+		app_err("pkt_pool not found\n");
+		return -1;
+	}
+
+	out_queue = odp_queue_lookup("ipsec-out");
+	if (cargs->schedule || cargs->poll) {
+		if (out_queue == ODP_QUEUE_INVALID) {
+			app_err("ipsec-out queue not found\n");
+			return -1;
+		}
+	}
+
+	if (cargs->reuse_packet) {
+		pkt = make_packet(pkt_pool, payload_length);
+		if (ODP_PACKET_INVALID == pkt)
+			return -1;
+	}
+
+	time_record_t start, end;
+	int packets_sent = 0;
+	int packets_received = 0;
+
+	/* Initialize parameters block */
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.num_opt = 0;
+	param.sa = &sa;
+
+	fill_time_record(&start);
+
+	while ((packets_sent < cargs->iteration_count) ||
```

Comment:
this function has to be spitted up to 2 or 3 small functions. with removing declaration variables in the middles. also  while cycles will be more easy to read.

Reply:
I've split this into sync and async cases. Splitting it further does not make  real sense, as it will lower performance.